jbmc/src/java_bytecode/java_object_factory.cpp

```diff
@@ -384,67 +385,95 @@ void java_object_factoryt::gen_pointer_target_init(
-  PRECONDITION(update_in_place!=update_in_placet::MAY_UPDATE_IN_PLACE);
+  PRECONDITION(expr.type().id() == ID_pointer);
+  PRECONDITION(update_in_place != update_in_placet::MAY_UPDATE_IN_PLACE);
 
-  if(target_type.id()==ID_struct &&
-     has_prefix(
-       id2string(to_struct_type(target_type).get_tag()),
-       "java::array["))
-  {
-    gen_nondet_array_init(
-      assignments,
-      expr,
-      depth+1,
-      update_in_place);
-  }
-  else
+  if(target_type.id() == ID_struct)
   {
-    // obtain a target pointer to initialize; if in MUST_UPDATE_IN_PLACE mode we
-    // initialize the fields of the object pointed by `expr`; if in
-    // NO_UPDATE_IN_PLACE then we allocate a new object, get a pointer to it
-    // (return value of `allocate_object`), emit a statement of the form
-    // `<expr> := address-of(<new-object>)` and recursively initialize such new
-    // object.
-    exprt target;
-    if(update_in_place==update_in_placet::NO_UPDATE_IN_PLACE)
+    const auto &target_class_type = to_java_class_type(target_type);
+    if(has_prefix(id2string(target_class_type.get_tag()), "java::array["))
     {
-      target=allocate_object(
-        assignments,
-        expr,
-        target_type,
-        alloc_type);
-      INVARIANT(
-        target.type().id()==ID_pointer,
-        "Pointer-typed expression expected");
+      gen_nondet_array_init(assignments, expr, depth + 1, update_in_place);
+      return;
     }
-    else
+    if(target_class_type.get_base("java::java.lang.Enum"))
     {
-      target=expr;
+      // We nondet-initialize enums to be equal to one of the constants defined
+      // for their type:
+      // int i = nondet(int)
+      // assume(0 < = i < $VALUES.length)
+      // expr = $VALUES[i]
+      // where $VALUES is a variable generated by the Java compiler that stores
+      // the array that is returned by Enum.values().
+
+      const irep_idt &class_name = target_class_type.get_name();
+      const symbolt &values = ns.lookup(id2string(class_name) + ".$VALUES");
+
+      // Access members (length and data) of $VALUES array
+      dereference_exprt deref_expr(values.symbol_expr());
+      const auto &deref_struct_type =
+        to_struct_type(ns.follow(deref_expr.type()));
+      const auto &comps = deref_struct_type.components();
+      const member_exprt length_expr(deref_expr, "length", comps[1].type());
+      const member_exprt enum_array_expr =
+        member_exprt(deref_expr, "data", comps[2].type());
+
+      const symbol_exprt &index_expr = gen_nondet_int_init(
+        assignments,
+        "enum_index_init",
+        from_integer(0, java_int_type()),
+        length_expr,
+        false);
+
+      // Index array using pointer arithmetic
+      plus_exprt plus(enum_array_expr, index_expr, enum_array_expr.type());
+      const dereference_exprt arraycellref(
+        plus_exprt(enum_array_expr, index_expr, enum_array_expr.type()),
+        enum_array_expr.type().subtype());
```

Comment:
In the case of a reference typed array, the type may need to be explicitly specified.

Reply:
Why? If it does, it suggests a type cast is needed, if the type if the `expr` passed in as the first parameter is T*, then any well formed `dereference_exprt` will have type `T` (i.e. `expr.type().subtype()`)

Reply:
@thk123 I removed the type. I checked the two constructors and the single parameter version does just call the other one with `op.type().subtype()`.

Reply:
Yeah the only way this could be different is if the `type()` of `plus` is different from `array_enum_expr`. The type of `plus` is determines by the lhs expr so in this case you're definitely fine. I think it would be valid to do `index_expr + enum_array_expr`  and then you would get the wrong answer...